docs/qos_enforcement.md

```diff
@@ -0,0 +1,143 @@
+  max_req_bw_ul: uint32
+  # Allocation and retention priority
+  arp:
+    arp_priority_level: uint32
+    arp_preemption_capability: bool
+    arp_preemption_vulnerability: bool
+ # Aggregated Maximum bit rate
+ # for non-GBR bearers, this limits the aggregate bit rate
+ # Enforce this on pipelined
+ apn_ambr:
+   apn_ambr_dl: uint32
+   apn_ambr_ul: uint32
+```
+
+As an initial goal, we would like to support apn_ambr.
+
+```
+apn_configuration:
+  supported_qos_profiles: array[qos_profile]
+```
+
+Multiple QoS objects will be allowed to be associated to APN objects.
+
+One can be specified as the APN-level default policy. This is checked by verifying a non-zero value for fields *apn_ambr_dl* and *apn_ambr_ul* for the default QoS object.
+
+Additional QoS objects can be associated to the APN for additional services beyond the default policy. For fields *apn_ambr_dl* and *apn_ambr_ul*, these field values must be zero.
+
+```
+subscriber:
+  id: string
+  name: string
+  ...
+  active_base_names: array[base_name]
+  active_policies: array[policy_id]
+  active_apns: array[string]
+  ...
+```
+
+One or more APN objects will be allowed to be associated to subscribers. This is similar to before, though will now create restrictions on associating policies to subscribers.
+
+As before, one or more policies will be allowed to be associated to subscribers. A new limitation will be created where each policy that is associated to a subscriber must share at least one of its QosProfile’s APNs with the subscriber. This must hold true for each policy to subscriber association, and CRUD operations which could break this rule should return an API error and fail to complete.
+
+```
+policy_rule:
+  id: string
+  flow_list: array[flow_description]
+  priority: uint32
+  rating_group: uint32
+  monitoring_key: string
+  ...
+  qos: array[qos_profile]
+```
+
+Multiple QoS objects will be allowed to be associated to each policy object. To implement this, its *qos* field will be changed from the current *flow_qos* type to the new QoS object type, as defined above.
+
+
+### Enforcement with Pipelined Service
+
+Changes in pipelined are required to enforce QoS via AMBR across the bearer level, APN level, and subscriber level.
+
+
+### Policydb Changes with Streaming
+
+The orc8r policydb service should be updated to stream to the gateway all the mappings between basenames, policies, QoS profiles, APNs, data plans and subscribers.
+
+Gateway policydb service needs to be updated to make RAR type calls to sessiond when updates happen to the policy associations or to policies. To know when this is required, policy information and associations are streamed from orc8r, and policydb should also query sessiond for the relevant info.
+
+
+### Session Manager Changes
+
+The session manager should provide grpc methods to retrieve the currently active sessions. With each session, the IMSI, APN, and active policies should be provided. This surface will be used by policydb.
+
```

Comment:
is this needed? sessiond already sends create/terminate session requests to policydb. So policydb can track which sessions are active and push any modifications on policy rules.

Reply:
Having `sessiond` be the source of truth here can keep `policydb` stateless. On principle also, I'm a strong believer of making sure we don't have duplicated information.